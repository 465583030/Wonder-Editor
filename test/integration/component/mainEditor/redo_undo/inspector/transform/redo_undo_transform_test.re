open Wonder_jest;

open Expect;

open Expect.Operators;

open Sinon;

let _ =
  describe(
    "redo_undo: transform",
    () => {
      let sandbox = getSandboxDefaultVal();
      let _buildMainEditorTransformComponent = (transformComponent) =>
        ReactTestRenderer.create(
          <MainEditorTransform
            store=(TestToolUI.buildEmptyAppState())
            dispatch=(TestToolUI.getDispatch())
            transformComponent
          />
        );
      beforeEach(
        () => {
          sandbox := createSandbox();
          TestToolEngine.prepare(sandbox);
          TestToolUI.initMainEditor(sandbox)
        }
      );
      afterEach(() => restoreSandbox(refJsObjToSandbox(sandbox^)));
      describe(
        "test set currentGameObject",
        () => {
          let _buildEngineSceneTree = () =>
            ReactTestRenderer.create(
              <MainEditorSceneTree
                store=(SceneTreeToolUI.buildAppStateSceneGraphFromEngine())
                dispatch=(TestToolUI.getDispatch())
              />
            );
          let _setSpecificGameObject = (clickTreeNodeIndex) => {
            let component = _buildEngineSceneTree();
            EventToolUI.triggerComponentEvent(
              component,
              SceneTreeEventUtils.triggerClickEvent(clickTreeNodeIndex)
            );
            MainEditorSceneToolEditor.unsafeGetCurrentGameObject() |> ignore
          };
          beforeEach(
            () => {
              TestToolEditor.closeContractCheck();
              StateHistoryToolEditor.clearAllState();
              MainEditorSceneToolEditor.prepareDefaultScene(
                MainEditorSceneToolEditor.setBoxTobeCurrentGameObject
              );
              _setSpecificGameObject(1)
            }
          );
          afterEach(() => TestToolEditor.openContractCheck());
          describe(
            "test snapshot",
            () => {
              describe(
                "undo operator",
                () => {
                  test(
                    "test no undo",
                    () => {
                      let currentGameObjectTransform =
                        MainEditorSceneToolEditor.getCurrentGameObjectTransform();
                      let component =
                        _buildMainEditorTransformComponent(currentGameObjectTransform);
                      EventToolUI.triggerComponentEvent(
                        component,
                        TransformEventUtils.triggerChangeXEvent("11.25")
                      );
                      EventToolUI.triggerComponentEvent(
                        component,
                        TransformEventUtils.triggerBlurXEvent("11.25")
                      );
                      EventToolUI.triggerComponentEvent(
                        component,
                        TransformEventUtils.triggerChangeYEvent("15")
                      );
                      EventToolUI.triggerComponentEvent(
                        component,
                        TransformEventUtils.triggerBlurYEvent("15")
                      );
                      let component2 =
                        _buildMainEditorTransformComponent(currentGameObjectTransform);
                      let json = ReactTestRenderer.toJSON(component2);
                      toMatchSnapshot(expect(json))
                    }
                  );
                  describe(
                    "undo one step",
                    () =>
                      test(
                        "undo step from second to first",
                        () => {
                          let currentGameObjectTransform =
                            MainEditorSceneToolEditor.getCurrentGameObjectTransform();
                          let component =
                            _buildMainEditorTransformComponent(currentGameObjectTransform);
                          EventToolUI.triggerComponentEvent(
                            component,
                            TransformEventUtils.triggerChangeXEvent("11.25")
                          );
                          EventToolUI.triggerComponentEvent(
                            component,
                            TransformEventUtils.triggerBlurXEvent("11.25")
                          );
                          EventToolUI.triggerComponentEvent(
                            component,
                            TransformEventUtils.triggerChangeYEvent("15")
                          );
                          EventToolUI.triggerComponentEvent(
                            component,
                            TransformEventUtils.triggerBlurYEvent("15")
                          );
                          StateHistoryToolEditor.undo();
                          let component2 =
                            _buildMainEditorTransformComponent(currentGameObjectTransform);
                          let json = ReactTestRenderer.toJSON(component2);
                          toMatchSnapshot(expect(json))
                        }
                      )
                  );
                  describe(
                    "undo two step",
                    () =>
                      test(
                        "undo step from second to zero",
                        () => {
                          let currentGameObjectTransform =
                            MainEditorSceneToolEditor.getCurrentGameObjectTransform();
                          let component =
                            _buildMainEditorTransformComponent(currentGameObjectTransform);
                          EventToolUI.triggerComponentEvent(
                            component,
                            TransformEventUtils.triggerChangeXEvent("11.25")
                          );
                          EventToolUI.triggerComponentEvent(
                            component,
                            TransformEventUtils.triggerBlurXEvent("11.25")
                          );
                          EventToolUI.triggerComponentEvent(
                            component,
                            TransformEventUtils.triggerChangeYEvent("15")
                          );
                          EventToolUI.triggerComponentEvent(
                            component,
                            TransformEventUtils.triggerBlurYEvent("15")
                          );
                          StateHistoryToolEditor.undo();
                          StateHistoryToolEditor.undo();
                          let component2 =
                            _buildMainEditorTransformComponent(currentGameObjectTransform);
                          let json = ReactTestRenderer.toJSON(component2);
                          toMatchSnapshot(expect(json))
                        }
                      )
                  )
                }
              );
              describe(
                "redo operator",
                () => {
                  describe(
                    "redo one step",
                    () =>
                      test(
                        "undo step from second to zero, redo step from zero to first",
                        () => {
                          let currentGameObjectTransform =
                            MainEditorSceneToolEditor.getCurrentGameObjectTransform();
                          let component =
                            _buildMainEditorTransformComponent(currentGameObjectTransform);
                          EventToolUI.triggerComponentEvent(
                            component,
                            TransformEventUtils.triggerChangeZEvent("1.2512")
                          );
                          EventToolUI.triggerComponentEvent(
                            component,
                            TransformEventUtils.triggerBlurZEvent("1.2512")
                          );
                          EventToolUI.triggerComponentEvent(
                            component,
                            TransformEventUtils.triggerChangeYEvent("-2.213")
                          );
                          EventToolUI.triggerComponentEvent(
                            component,
                            TransformEventUtils.triggerBlurYEvent("-2.213")
                          );
                          StateHistoryToolEditor.undo();
                          StateHistoryToolEditor.undo();
                          StateHistoryToolEditor.redo();
                          let component2 =
                            _buildMainEditorTransformComponent(currentGameObjectTransform);
                          let json = ReactTestRenderer.toJSON(component2);
                          toMatchSnapshot(expect(json))
                        }
                      )
                  );
                  describe(
                    "redo two step",
                    () =>
                      test(
                        "undo step from second to zero, redo step from zero to second",
                        () => {
                          let currentGameObjectTransform =
                            MainEditorSceneToolEditor.getCurrentGameObjectTransform();
                          let component =
                            _buildMainEditorTransformComponent(currentGameObjectTransform);
                          EventToolUI.triggerComponentEvent(
                            component,
                            TransformEventUtils.triggerChangeZEvent("25.11")
                          );
                          EventToolUI.triggerComponentEvent(
                            component,
                            TransformEventUtils.triggerBlurZEvent("25.11")
                          );
                          EventToolUI.triggerComponentEvent(
                            component,
                            TransformEventUtils.triggerChangeYEvent("15.12")
                          );
                          EventToolUI.triggerComponentEvent(
                            component,
                            TransformEventUtils.triggerBlurYEvent("15.12")
                          );
                          StateHistoryToolEditor.undo();
                          StateHistoryToolEditor.undo();
                          StateHistoryToolEditor.redo();
                          StateHistoryToolEditor.redo();
                          let component2 =
                            _buildMainEditorTransformComponent(currentGameObjectTransform);
                          let json = ReactTestRenderer.toJSON(component2);
                          toMatchSnapshot(expect(json))
                        }
                      )
                  );
                  describe(
                    "redo three step",
                    () =>
                      test(
                        "undo step from second to zero, redo step from zero to thrid",
                        () => {
                          let currentGameObjectTransform =
                            MainEditorSceneToolEditor.getCurrentGameObjectTransform();
                          let component =
                            _buildMainEditorTransformComponent(currentGameObjectTransform);
                          EventToolUI.triggerComponentEvent(
                            component,
                            TransformEventUtils.triggerChangeZEvent("25.11")
                          );
                          EventToolUI.triggerComponentEvent(
                            component,
                            TransformEventUtils.triggerBlurZEvent("25.11")
                          );
                          EventToolUI.triggerComponentEvent(
                            component,
                            TransformEventUtils.triggerChangeYEvent("15.12")
                          );
                          EventToolUI.triggerComponentEvent(
                            component,
                            TransformEventUtils.triggerBlurYEvent("15.12")
                          );
                          StateHistoryToolEditor.undo();
                          StateHistoryToolEditor.undo();
                          StateHistoryToolEditor.redo();
                          StateHistoryToolEditor.redo();
                          StateHistoryToolEditor.redo();
                          let component2 =
                            _buildMainEditorTransformComponent(currentGameObjectTransform);
                          let json = ReactTestRenderer.toJSON(component2);
                          toMatchSnapshot(expect(json))
                        }
                      )
                  )
                }
              )
            }
          )
        }
      )
    }
  );