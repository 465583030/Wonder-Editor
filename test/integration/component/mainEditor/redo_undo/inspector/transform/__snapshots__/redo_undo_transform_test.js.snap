// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redo_undo: transform test set currentGameObject test snapshot redo operator redo one step undo step from second to zero, redo step from zero to first 1`] = `
<article
  className="transform-component"
>
  <article
    className="wonder-float-input"
  >
    <span
      className="component-label"
    >
      X : 
    </span>
    <input
      className="input-component float-input"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
      value="0"
    />
  </article>
  <article
    className="wonder-float-input"
  >
    <span
      className="component-label"
    >
      Y : 
    </span>
    <input
      className="input-component float-input"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
      value="0"
    />
  </article>
  <article
    className="wonder-float-input"
  >
    <span
      className="component-label"
    >
      Z : 
    </span>
    <input
      className="input-component float-input"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
      value="1.2512"
    />
  </article>
</article>
`;

exports[`redo_undo: transform test set currentGameObject test snapshot redo operator redo three step undo step from second to zero, redo step from zero to thrid 1`] = `
<article
  className="transform-component"
>
  <article
    className="wonder-float-input"
  >
    <span
      className="component-label"
    >
      X : 
    </span>
    <input
      className="input-component float-input"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
      value="0"
    />
  </article>
  <article
    className="wonder-float-input"
  >
    <span
      className="component-label"
    >
      Y : 
    </span>
    <input
      className="input-component float-input"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
      value="15.12"
    />
  </article>
  <article
    className="wonder-float-input"
  >
    <span
      className="component-label"
    >
      Z : 
    </span>
    <input
      className="input-component float-input"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
      value="25.110001"
    />
  </article>
</article>
`;

exports[`redo_undo: transform test set currentGameObject test snapshot redo operator redo two step undo step from second to zero, redo step from zero to second 1`] = `
<article
  className="transform-component"
>
  <article
    className="wonder-float-input"
  >
    <span
      className="component-label"
    >
      X : 
    </span>
    <input
      className="input-component float-input"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
      value="0"
    />
  </article>
  <article
    className="wonder-float-input"
  >
    <span
      className="component-label"
    >
      Y : 
    </span>
    <input
      className="input-component float-input"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
      value="15.12"
    />
  </article>
  <article
    className="wonder-float-input"
  >
    <span
      className="component-label"
    >
      Z : 
    </span>
    <input
      className="input-component float-input"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
      value="25.110001"
    />
  </article>
</article>
`;

exports[`redo_undo: transform test set currentGameObject test snapshot undo operator test no undo 1`] = `
<article
  className="transform-component"
>
  <article
    className="wonder-float-input"
  >
    <span
      className="component-label"
    >
      X : 
    </span>
    <input
      className="input-component float-input"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
      value="11.25"
    />
  </article>
  <article
    className="wonder-float-input"
  >
    <span
      className="component-label"
    >
      Y : 
    </span>
    <input
      className="input-component float-input"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
      value="15"
    />
  </article>
  <article
    className="wonder-float-input"
  >
    <span
      className="component-label"
    >
      Z : 
    </span>
    <input
      className="input-component float-input"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
      value="0"
    />
  </article>
</article>
`;

exports[`redo_undo: transform test set currentGameObject test snapshot undo operator undo one step undo step from second to first 1`] = `
<article
  className="transform-component"
>
  <article
    className="wonder-float-input"
  >
    <span
      className="component-label"
    >
      X : 
    </span>
    <input
      className="input-component float-input"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
      value="11.25"
    />
  </article>
  <article
    className="wonder-float-input"
  >
    <span
      className="component-label"
    >
      Y : 
    </span>
    <input
      className="input-component float-input"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
      value="0"
    />
  </article>
  <article
    className="wonder-float-input"
  >
    <span
      className="component-label"
    >
      Z : 
    </span>
    <input
      className="input-component float-input"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
      value="0"
    />
  </article>
</article>
`;

exports[`redo_undo: transform test set currentGameObject test snapshot undo operator undo two step undo step from second to zero 1`] = `
<article
  className="transform-component"
>
  <article
    className="wonder-float-input"
  >
    <span
      className="component-label"
    >
      X : 
    </span>
    <input
      className="input-component float-input"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
      value="0"
    />
  </article>
  <article
    className="wonder-float-input"
  >
    <span
      className="component-label"
    >
      Y : 
    </span>
    <input
      className="input-component float-input"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
      value="0"
    />
  </article>
  <article
    className="wonder-float-input"
  >
    <span
      className="component-label"
    >
      Z : 
    </span>
    <input
      className="input-component float-input"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
      value="0"
    />
  </article>
</article>
`;
